### ISSUE TITLE
[Done] SOLO: Segmenting Objects by Locations

### ISSUE THREAD

**Issue by phalanx-hk** [phalanx-hk/eccv2020_paperlist#2]
# Overview
SOLO is single shot instance segmentation approach. 
it reformulate two-stage instance segmentation(i.e., mask rcnn) as two sub tasks, predicting semantic category in category branch and masks of instances in mask branch. 
First, input image divided into SxS grid. Category branch predict semantic class probability for each grid and mask branch predict instance mask corresponding each grid. Finally, the category prediction and the corresponding mask are naturally associated by their reference grid cell.
![image](https://user-images.githubusercontent.com/44429453/91384366-54321580-e869-11ea-820b-6250f6a40ff8.png)

# Method
Input image is divided into SxS grids. If the center of an object fall into a grid cell, that grid cell is responsible for predicting semantic category and masks of instances.
## Sementic category
We predict semantic class probabilities. output shape is CxSxS, C is number of category. Each cell of the SxS grid must belong to one individual instance. During inference, C-dimentional output indicates the class probability for each object instance.
- label assignment and loss function
Given the ground truth mask, mask center(cx, cy) is obtained by the same center sampling method as FCOS. Center region is defined (cx, cy, εw, εh), where w/h are mask width and height and ε is scale factor(0.2 in journal).
Loss function is conventional focal loss.

## Instance mask
We predict instance mask corresponding each grid, so there are S^2 predicted mask in total. We encode these masks at the third dimension of 3D output tensor. Output shape is HxWxS^2 where kth channel is responsible to segment instance at grid.
Finally, one-to-one correspondence is established between the semantic category and class-agnostic mask.
Instance mask branch need to correspond position information in image since segmentation masks are conditioned on the grid cells and must be separated by different feature channels. So we use CoordConv for corresponding position information.
- label assignment and loss function

## Inference
We use a confidence threshold of 0.1 to filter out category score with low confidence. Then we select the top 500 scoring masks and feed them into the NMS. Mask threshold is 0.5.
Also, we calculate maskness for each predicted mask, which represents the quality and confidence of mask prediction maskness. Maskness is average of predicted soft mask value. The classification score for each prdiction is multiplied by the maskness as the final confidence score.

# Result
## compare mask AP with SOTA method
use coco test-dev dataset
SOLO outperforms all previous one-stage methods
![image](https://user-images.githubusercontent.com/44429453/91416310-2d380b80-e88a-11ea-8d6a-3352b8512aa1.png)

## impact of grid number and FPN
We compare the impact of grid number on the performance with single output feature map. 
Feature is generated by merging C3, C4, C5.
S = 12 can acieve 27.2 AP -> single-scale SOLO can be applicable to some scenarios where object scales don't vary much.
Also, experiment multi-level prediction using FPN.  From P2 to P6, the corresponding grid numbers are [40, 36, 24, 16, 12] respectively. It achieve 35.8 AP.
![image](https://user-images.githubusercontent.com/44429453/91416685-b7806f80-e88a-11ea-8521-262a19368dc0.png)

## CoordConv
Adding single CoordConv improve AP(+3.6). Two or more CoordConvs don't bring noticeable improvement.
![image](https://user-images.githubusercontent.com/44429453/91417489-d6cbcc80-e88b-11ea-9f55-2f604f6c54d1.png)

## Loss function
compare different mask loss functions. 
Focal Loss is better than BCE because many pixels of intace mask are in background.
Dice Loss is better than Focal Loss because it views the pixels as a whole object and could establish the right balance between foreground and background pixels automatically.
BCE and Focal Loss may be better than Dice Loss by hyper-parameter tuning, but Dice Loss is stable and more likely to attain good results without heuristics.
![image](https://user-images.githubusercontent.com/44429453/91418401-0b8c5380-e88d-11ea-8acb-ee964c07779f.png)

## Alignment in the category branch
In the category branch, we must must match the convolutional features with spatial size H×W to S×S.
![image](https://user-images.githubusercontent.com/44429453/91418575-49897780-e88d-11ea-8ce6-bf138931ba49.png)
So we compare three alignment methods, directly bilinear interpolation, avgpool, region grid interpolation.
From our observation, there is no noticeable performance gap between these variants.

## SOLO-512
use shorter image size of 512 instead of 800.
Models run on single V100 GPU.
![image](https://user-images.githubusercontent.com/44429453/91418958-cae10a00-e88d-11ea-8968-7ef8b694501b.png)

## Error Analysis
We perform an error analysis by replacing the predicted masks with ground-truth values.
For each predicted binary mask, we compute IoUs with ground-truth masks, and replace it with the most overlapping ground-truth mask.
Replacing predicted masks with ground truth masks increase AP to 68.1%, so  there are still ample room for improving the mask branch.
![image](https://user-images.githubusercontent.com/44429453/91419145-08de2e00-e88e-11ea-9531-82a241bbf8aa.png)

## Decoupled SOLO
we set grid number S = 20, output is 400 channel maps. that prediction is somewhat redundant as in most cases the objects are located sparsely in the image. So we propose Decoupled head.
As below image, We replace output tensor with two output tensors. Each output tensor correspond two axes respectively. Thus, output space is decreased from HxWxS^2 to HxWx2S. 
For an object located at grid location (i, j), the mask prediction of that object is defined as the element-wise multiplication of two channel maps.
![image](https://user-images.githubusercontent.com/44429453/91419880-e13b9580-e88e-11ea-8f86-7ae0fc3bd8b4.png)
Decoupled SOLO improve AP to 38.4, so it is efficient and equivalent variant in accuracy of SOLO.
![image](https://user-images.githubusercontent.com/44429453/91419765-c36e3080-e88e-11ea-9192-8ee9ddf487b5.png)
